### PR TITLE
Cloud spanner templates, tests and doco

### PR DESCRIPTION
**Pull request by sourced-nan**
Cloud spanner templates, tests and doco

### CODE REVIEW

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py [hunk 1]
```diff
@@ -26,12 +24,11 @@ def get_spanner_instance_config(project_id, config):
...
-    Input: context
-           context is generated by gcloud when loading the input config file
-    Returns: dictionary with key name resources
-           This contains all the information gcloud needs to create a spanner
-           instance, databases and permissions.
+    Generates the config gcloud needs to create a Cloud Spanner instance.
+    Input: context - generated by gcloud when loading the input config file.
+    Output: dictionary with key resource names - this contains all the 
+            information gcloud needs to create a spanner instance, 
+            databases ??? "databases" plural ??? and permissions.
```
Comment (sourced-vince) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: yes. Can be multiple databases. Plural.

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py.schema [hunk 2]
```diff
@@ -59,31 +59,31 @@ properties:
...
-          description: Role been assigned as part of IAM policy
+          description: The IAM role assigned as part of the IAM policy.
           enum:
             - roles/spanner.admin
             - roles/spanner.viewer
             - roles/spanner.databaseAdmin
         members:
           type: array
-          description: Members been granted as given role for the IAM policy
+          description: A list of members granted the given role for ??? under ??? the IAM policy.
```
Comment (sourced-vince) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py.schema]: for

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py.schema [hunk 3]
```diff
@@ -92,14 +92,14 @@ properties:
             properties:
               role:
                 type: string
-                description: Role been assigned as part of IAM policy
+                description: The role assigned as part of ??? under ??? the IAM policy.
```
Comment (sourced-vince) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py.schema]: for

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py.schema [hunk 4]
```diff
@@ -92,14 +92,14 @@ properties:
...
                 enum:
                   - roles/spanner.databaseAdmin
                   - roles/spanner.databaseReader
                   - roles/spanner.databaseUser
               members:
                 type: array
-                description: Members been granted as given role for the IAM policy
+                description: A list of members granted the given role for ??? under ??? the IAM policy.
```
Comment (sourced-vince) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py.schema]: for

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" This template creates a Cloud Spanner instance and database. """
+
+
+def get_spanner_instance_id(project_id, base_name):
+    ''' Generate the instance URL '''
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: Here and after, inconsistent docstring quotes: double first (also used in every other template), then single quotes here and after.

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    resources_list.append(resource)
+
+    if context.properties.get('bindings'):
+        policy = {
+            'name':
+                "{}{}".format(context.env['name'],
+                              '-setIamPolicy'),
+            'action':
+                (
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: Why parentheses? As soon as they are gone and indentation fixed, `pylint` can be removed as well.

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        }
+        resources_list.append(database_resource)
+
+        if database.get('bindings'):
+            database_policy = {
+                'name':
+                    "{}{}".format(database_resource_name,
+                                  "-setIamPolicy"),
+                'action':
+                    (
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: Possibly redundant parentheses, indentation and pylint directive.

File: community/cloud-foundation/templates/cloud_spanner/examples/cloud_spanner.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# Example of the Cloud Spanner template usage.
+
+imports:
+- path: templates/cloud_spanner/cloud_spanner.py
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/examples/cloud_spanner.yaml]: Inconsistent indentation. Also a few places below

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+function teardown() {
+    if [[ "$BATS_TEST_NUMBER" -eq "${#BATS_TEST_NAMES[@]}" ]]; then
+        rm -f "${RANDOM_FILE}"
+        delete_config
+    fi
+}
+
+@test "Creating deployment ${DEPLOYMENT_NAME} from ${CONFIG}" {
+    gcloud deployment-manager deployments create "${DEPLOYMENT_NAME}" \
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Missing `run` before `gcloud`

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        --config ${CONFIG} \
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}"
+}
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Please add `[[ "$status" -eq 0 ]]` assertion here and as a first check in every other test

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+@test "Verifying that Spanner DB IAM was created as part of ${DEPLOYMENT_NAME}" {
+    run gcloud spanner databases get-iam-policy spannerdb1 --instance test-myspannercluster-"${RAND}" \
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}"
+    [[ "$output" =~ "role: roles/spanner.databaseAdmin" ]]
+}
+
+@test "Deleting deployment" {
+    gcloud deployment-manager deployments delete "${DEPLOYMENT_NAME}" \
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Add `run`

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}" -q
+    run gcloud  run gcloud spanner instances list \
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: I think this part is redundant. Our job ends as soon as deployment deletion is finished, so simply checking if `deployments delete` succeeded  (`[[ "$status" -eq 0 ]]`) should be enough. After all, if deployment deletion succeeds, but the instance is still there, what it is that we can do now?

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py (hunk 5 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: Fixed

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py (hunk 6 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: Fixed

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py (hunk 7 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: Fixed

File: community/cloud-foundation/templates/cloud_spanner/examples/cloud_spanner.yaml (hunk 8 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/examples/cloud_spanner.yaml]: Fixed

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats (hunk 9 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Fixed

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats (hunk 10 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Fixed

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats (hunk 11 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Fixed

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats (hunk 12 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Will leave in to ensure resource is removed.

File: community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+
+def get_spanner_instance_config(project_id, config):
+    """ Generate the instance config URL """
+
+    return "projects/{}/instanceConfigs/{}".format(project_id, config)
+
+
+def generate_config(context):
+    '''
```
Comment (sourced-pavel) [inline community/cloud-foundation/templates/cloud_spanner/cloud_spanner.py]: One more singlequoted docstring

**Pull request closed without merging**